Sharing violation on loading registry.json

### Is there an existing issue for this?

- [x] I have checked the existing issues for an existing issue. I even actually looked at the ones that are pinned to the top!

### Operating System

macOS 12.6.3

### CKAN Version

ckan v1.35.0.24270

### Games

KSP 1

### Game Version

1.12.5

### Did you make any manual changes to your game folder (i.e., not via CKAN)?

_No response_

### Describe the bug

when i tried to add a new ksp folder for modding mono crashes and shows a very long error message

### Steps to reproduce

launch ckan consoleui
add ksp folder
mono crashes
error message shows

### Relevant log output

```
172408 [1] ERROR CKAN.RegistryManager (null) - Uncaught exception loading registry: System.IO.IOException: Sharing violation on path /volumes/tf-256/ksp files/ksp modded/CKAN/registry.json
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) [0x0019e] in <207d12884485488d9288cc8cbf474289>:0 
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.IO.FileOptions options) [0x00000] in <207d12884485488d9288cc8cbf474289>:0 
  at (wrapper remoting-invoke-with-check) System.IO.FileStream..ctor(string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare,int,System.IO.FileOptions)
  at System.IO.StreamReader..ctor (System.String path, System.Text.Encoding encoding, System.Boolean detectEncodingFromByteOrderMarks, System.Int32 bufferSize) [0x00055] in <207d12884485488d9288cc8cbf474289>:0 
  at System.IO.StreamReader..ctor (System.String path, System.Text.Encoding encoding, System.Boolean detectEncodingFromByteOrderMarks) [0x00000] in <207d12884485488d9288cc8cbf474289>:0 
  at (wrapper remoting-invoke-with-check) System.IO.StreamReader..ctor(string,System.Text.Encoding,bool)
  at System.IO.File.InternalReadAllText (System.String path, System.Text.Encoding encoding) [0x00000] in <207d12884485488d9288cc8cbf474289>:0 
  at System.IO.File.ReadAllText (System.String path) [0x0002c] in <207d12884485488d9288cc8cbf474289>:0 
  at CKAN.RegistryManager.Load (CKAN.RepositoryDataManager repoData) [0x00035] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.RegistryManager.LoadOrCreate (CKAN.RepositoryDataManager repoData) [0x00000] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 

Unhandled Exception:
System.IO.IOException: Sharing violation on path /volumes/tf-256/ksp files/ksp modded/CKAN/registry.json
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) [0x0019e] in <207d12884485488d9288cc8cbf474289>:0 
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.IO.FileOptions options) [0x00000] in <207d12884485488d9288cc8cbf474289>:0 
  at (wrapper remoting-invoke-with-check) System.IO.FileStream..ctor(string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare,int,System.IO.FileOptions)
  at System.IO.StreamReader..ctor (System.String path, System.Text.Encoding encoding, System.Boolean detectEncodingFromByteOrderMarks, System.Int32 bufferSize) [0x00055] in <207d12884485488d9288cc8cbf474289>:0 
  at System.IO.StreamReader..ctor (System.String path, System.Text.Encoding encoding, System.Boolean detectEncodingFromByteOrderMarks) [0x00000] in <207d12884485488d9288cc8cbf474289>:0 
  at (wrapper remoting-invoke-with-check) System.IO.StreamReader..ctor(string,System.Text.Encoding,bool)
  at System.IO.File.InternalReadAllText (System.String path, System.Text.Encoding encoding) [0x00000] in <207d12884485488d9288cc8cbf474289>:0 
  at System.IO.File.ReadAllText (System.String path) [0x0002c] in <207d12884485488d9288cc8cbf474289>:0 
  at CKAN.RegistryManager.Load (CKAN.RepositoryDataManager repoData) [0x00035] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.RegistryManager.LoadOrCreate (CKAN.RepositoryDataManager repoData) [0x000be] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.RegistryManager..ctor (System.String path, CKAN.GameInstance inst, CKAN.RepositoryDataManager repoData) [0x00067] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.RegistryManager.Instance (CKAN.GameInstance inst, CKAN.RepositoryDataManager repoData) [0x00024] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.ConsoleUI.ConsoleCKAN..ctor (CKAN.GameInstanceManager mgr, System.String themeName, System.String userAgent, System.Boolean debug) [0x0009b] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.ConsoleUI.ConsoleUI.Main_ (CKAN.GameInstanceManager manager, System.String themeName, System.String userAgent, System.Boolean debug) [0x00005] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.CmdLine.MainClass.ConsoleUi (CKAN.GameInstanceManager manager, CKAN.CmdLine.ConsoleUIOptions opts) [0x00039] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.CmdLine.MainClass.RunSimpleAction (CKAN.CmdLine.Options cmdline, CKAN.CmdLine.CommonOptions options, System.String[] args, CKAN.IUser user, CKAN.GameInstanceManager manager) [0x002aa] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.CmdLine.MainClass.Execute (CKAN.GameInstanceManager manager, CKAN.CmdLine.CommonOptions opts, System.String[] args) [0x002b9] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.CmdLine.MainClass.Main (System.String[] args) [0x000e9] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.IO.IOException: Sharing violation on path /volumes/tf-256/ksp files/ksp modded/CKAN/registry.json
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) [0x0019e] in <207d12884485488d9288cc8cbf474289>:0 
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.IO.FileOptions options) [0x00000] in <207d12884485488d9288cc8cbf474289>:0 
  at (wrapper remoting-invoke-with-check) System.IO.FileStream..ctor(string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare,int,System.IO.FileOptions)
  at System.IO.StreamReader..ctor (System.String path, System.Text.Encoding encoding, System.Boolean detectEncodingFromByteOrderMarks, System.Int32 bufferSize) [0x00055] in <207d12884485488d9288cc8cbf474289>:0 
  at System.IO.StreamReader..ctor (System.String path, System.Text.Encoding encoding, System.Boolean detectEncodingFromByteOrderMarks) [0x00000] in <207d12884485488d9288cc8cbf474289>:0 
  at (wrapper remoting-invoke-with-check) System.IO.StreamReader..ctor(string,System.Text.Encoding,bool)
  at System.IO.File.InternalReadAllText (System.String path, System.Text.Encoding encoding) [0x00000] in <207d12884485488d9288cc8cbf474289>:0 
  at System.IO.File.ReadAllText (System.String path) [0x0002c] in <207d12884485488d9288cc8cbf474289>:0 
  at CKAN.RegistryManager.Load (CKAN.RepositoryDataManager repoData) [0x00035] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.RegistryManager.LoadOrCreate (CKAN.RepositoryDataManager repoData) [0x000be] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.RegistryManager..ctor (System.String path, CKAN.GameInstance inst, CKAN.RepositoryDataManager repoData) [0x00067] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.RegistryManager.Instance (CKAN.GameInstance inst, CKAN.RepositoryDataManager repoData) [0x00024] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.ConsoleUI.ConsoleCKAN..ctor (CKAN.GameInstanceManager mgr, System.String themeName, System.String userAgent, System.Boolean debug) [0x0009b] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.ConsoleUI.ConsoleUI.Main_ (CKAN.GameInstanceManager manager, System.String themeName, System.String userAgent, System.Boolean debug) [0x00005] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.CmdLine.MainClass.ConsoleUi (CKAN.GameInstanceManager manager, CKAN.CmdLine.ConsoleUIOptions opts) [0x00039] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.CmdLine.MainClass.RunSimpleAction (CKAN.CmdLine.Options cmdline, CKAN.CmdLine.CommonOptions options, System.String[] args, CKAN.IUser user, CKAN.GameInstanceManager manager) [0x002aa] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.CmdLine.MainClass.Execute (CKAN.GameInstanceManager manager, CKAN.CmdLine.CommonOptions opts, System.String[] args) [0x002b9] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0 
  at CKAN.CmdLine.MainClass.Main (System.String[] args) [0x000e9] in <e7a11ea66e094dd1a3ac6d97032c8cd9>:0
```

--- a/Core/Registry/Registry.cs
+++ b/Core/Registry/Registry.cs
@@ -374,8 +374,8 @@ namespace CKAN
         {
         }
 
-        public static Registry Empty()
-            => new Registry(null,
+        public static Registry Empty(RepositoryDataManager repoData)
+            => new Registry(repoData,
                             new Dictionary<string, InstalledModule>(),
                             new Dictionary<string, string>(),
                             new Dictionary<string, string>(),

--- a/Core/Registry/RegistryManager.cs
+++ b/Core/Registry/RegistryManager.cs
@@ -70,7 +70,7 @@ namespace CKAN
 
             try
             {
-                LoadOrCreate(repoData);
+                Load(repoData);
             }
             catch
             {
@@ -307,40 +307,29 @@ namespace CKAN
         [MemberNotNull(nameof(registry))]
         private void Load(RepositoryDataManager repoData)
         {
-            // Our registry needs to know our game instance when upgrading from older
-            // registry formats. This lets us encapsulate that to make it available
-            // after deserialisation.
-            var settings = new JsonSerializerSettings
-            {
-                DateTimeZoneHandling = DateTimeZoneHandling.Utc,
-                Context = new StreamingContext(StreamingContextStates.Other, gameInstance)
-            };
-
-            log.DebugFormat("Trying to load registry from {0}", path);
-            string json = File.ReadAllText(path);
-            log.Debug("Registry JSON loaded; parsing...");
-            registry = new Registry(repoData);
-            JsonConvert.PopulateObject(json, registry, settings);
-            log.Debug("Registry loaded and parsed");
-            log.InfoFormat("Loaded CKAN registry at {0}", path);
-        }
-
-        [MemberNotNull(nameof(registry))]
-        private void LoadOrCreate(RepositoryDataManager repoData)
-        {
             try
             {
-                Load(repoData);
+                // Our registry needs to know our game instance when upgrading from older
+                // registry formats. This lets us encapsulate that to make it available
+                // after deserialisation.
+                var settings = new JsonSerializerSettings
+                {
+                    DateTimeZoneHandling = DateTimeZoneHandling.Utc,
+                    Context = new StreamingContext(StreamingContextStates.Other, gameInstance)
+                };
+
+                log.DebugFormat("Trying to load registry from {0}", path);
+                string json = File.ReadAllText(path);
+                log.Debug("Registry JSON loaded; parsing...");
+                registry = new Registry(repoData);
+                JsonConvert.PopulateObject(json, registry, settings);
+                log.Debug("Registry loaded and parsed");
+                log.InfoFormat("Loaded CKAN registry at {0}", path);
+                AscertainDefaultRepo();
             }
-            catch (FileNotFoundException)
+            catch (IOException exc) when (exc is FileNotFoundException or DirectoryNotFoundException)
             {
-                Create();
-                Load(repoData);
-            }
-            catch (DirectoryNotFoundException)
-            {
-                Create();
-                Load(repoData);
+                Create(repoData);
             }
             catch (JsonException exc)
             {
@@ -349,25 +338,22 @@ namespace CKAN
                 log.ErrorFormat("{0} is corrupted, archiving to {1}: {2}",
                     path, previousCorruptedPath, previousCorruptedMessage);
                 File.Move(path, previousCorruptedPath);
-                Create();
-                Load(repoData);
+                Create(repoData);
             }
             catch (Exception ex)
             {
                 log.ErrorFormat("Uncaught exception loading registry: {0}", ex.ToString());
                 throw;
             }
-            AscertainDefaultRepo();
         }
 
         [MemberNotNull(nameof(registry))]
-        private void Create()
+        private void Create(RepositoryDataManager repoData)
         {
             log.InfoFormat("Creating new CKAN registry at {0}", path);
-            registry = Registry.Empty();
+            registry = Registry.Empty(repoData);
             AscertainDefaultRepo();
             ScanUnmanagedFiles();
-            Save();
         }
 
         private void AscertainDefaultRepo()

--- a/Tests/Core/ModuleInstallerDirTest.cs
+++ b/Tests/Core/ModuleInstallerDirTest.cs
@@ -16,7 +16,7 @@ namespace Tests.Core
     /// Tests the AddParentDirectories method in CKAN.ModuleInstaller
     /// </summary>
     [TestFixture]
-    public class ModuleInstallerDirTest
+    public class ModuleInstallerDirTests
     {
         private GameInstanceManager? _manager;
         private DisposableKSP?       _instance;

--- a/Tests/Core/Net/NetFileCacheTests.cs
+++ b/Tests/Core/Net/NetFileCacheTests.cs
@@ -160,33 +160,41 @@ namespace Tests.Core
         [Test]
         public void EnforceSizeLimit_UnderLimit_FileRetained()
         {
-            // Arrange
-            CKAN.Registry registry = CKAN.Registry.Empty();
-            long fileSize = new FileInfo(TestData.DogeCoinFlagZip()).Length;
+            var user = new NullUser();
+            using (var repoData = new TemporaryRepositoryData(user))
+            {
+                // Arrange
+                CKAN.Registry registry = CKAN.Registry.Empty(repoData.Manager);
+                long fileSize = new FileInfo(TestData.DogeCoinFlagZip()).Length;
 
-            // Act
-            Uri url = new Uri("http://kitte.nz/");
-            cache?.Store(url, TestData.DogeCoinFlagZip());
-            cache?.EnforceSizeLimit(fileSize + 100, registry);
+                // Act
+                Uri url = new Uri("http://kitte.nz/");
+                cache?.Store(url, TestData.DogeCoinFlagZip());
+                cache?.EnforceSizeLimit(fileSize + 100, registry);
 
-            // Assert
-            Assert.IsTrue(cache?.IsCached(url));
+                // Assert
+                Assert.IsTrue(cache?.IsCached(url));
+            }
         }
 
         [Test]
         public void EnforceSizeLimit_OverLimit_FileRemoved()
         {
-            // Arrange
-            CKAN.Registry registry = CKAN.Registry.Empty();
-            long fileSize = new FileInfo(TestData.DogeCoinFlagZip()).Length;
+            var user = new NullUser();
+            using (var repoData = new TemporaryRepositoryData(user))
+            {
+                // Arrange
+                CKAN.Registry registry = CKAN.Registry.Empty(repoData.Manager);
+                long fileSize = new FileInfo(TestData.DogeCoinFlagZip()).Length;
 
-            // Act
-            Uri url = new Uri("http://kitte.nz/");
-            cache?.Store(url, TestData.DogeCoinFlagZip());
-            cache?.EnforceSizeLimit(fileSize - 100, registry);
+                // Act
+                Uri url = new Uri("http://kitte.nz/");
+                cache?.Store(url, TestData.DogeCoinFlagZip());
+                cache?.EnforceSizeLimit(fileSize - 100, registry);
 
-            // Assert
-            Assert.IsFalse(cache?.IsCached(url));
+                // Assert
+                Assert.IsFalse(cache?.IsCached(url));
+            }
         }
 
         #pragma warning disable IDE0051

--- a/Tests/Core/Registry/Registry.cs
+++ b/Tests/Core/Registry/Registry.cs
@@ -40,8 +40,12 @@ namespace Tests.Core.Registry
         [Test]
         public void Empty()
         {
-            CKAN.Registry registry = CKAN.Registry.Empty();
-            Assert.IsInstanceOf<CKAN.Registry>(registry);
+            var user = new NullUser();
+            using (var repoData = new TemporaryRepositoryData(user))
+            {
+                CKAN.Registry registry = CKAN.Registry.Empty(repoData.Manager);
+                Assert.IsInstanceOf<CKAN.Registry>(registry);
+            }
         }
 
         [Test]
@@ -434,10 +438,12 @@ namespace Tests.Core.Registry
                          }");
             CKAN.Registry reg;
 
+            var user = new NullUser();
+            using (var repoData = new TemporaryRepositoryData(user))
             using (var gameInstWrapper = new DisposableKSP())
             using (var tScope = new TransactionScope())
             {
-                reg = CKAN.Registry.Empty();
+                reg = CKAN.Registry.Empty(repoData.Manager);
                 reg.RegisterModule(module, new List<string>(),
                                    gameInstWrapper.KSP, false);
 
@@ -455,147 +461,163 @@ namespace Tests.Core.Registry
         [Test]
         public void TxCommit()
         {
-            // Our registry should work fine on committed transactions.
-            // This one seemingly just makes sure adding a mod adds it
-            // when the registry is created outside the transaction
-            var module = CkanModule.FromJson(@"{
-                             ""spec_version"": ""v1.4"",
-                             ""identifier"":   ""InstalledMod"",
-                             ""author"":       ""InstalledModder"",
-                             ""version"":      ""1.0"",
-                             ""download"":     ""https://github.com/""
-                         }");
-            var registry = CKAN.Registry.Empty();
-
-            using (var gameInstWrapper = new DisposableKSP())
-            using (var tScope = new TransactionScope())
+            var user = new NullUser();
+            using (var repoData = new TemporaryRepositoryData(user))
             {
-                registry.RegisterModule(module, new List<string>(),
-                                        gameInstWrapper.KSP, false);
+                // Our registry should work fine on committed transactions.
+                // This one seemingly just makes sure adding a mod adds it
+                // when the registry is created outside the transaction
+                var module = CkanModule.FromJson(@"{
+                                 ""spec_version"": ""v1.4"",
+                                 ""identifier"":   ""InstalledMod"",
+                                 ""author"":       ""InstalledModder"",
+                                 ""version"":      ""1.0"",
+                                 ""download"":     ""https://github.com/""
+                             }");
+                var registry = CKAN.Registry.Empty(repoData.Manager);
 
+                using (var gameInstWrapper = new DisposableKSP())
+                using (var tScope = new TransactionScope())
+                {
+                    registry.RegisterModule(module, new List<string>(),
+                                            gameInstWrapper.KSP, false);
+
+                    CollectionAssert.AreEqual(
+                        Enumerable.Repeat(module, 1),
+                        registry.InstalledModules.Select(im => im.Module));
+
+                    tScope.Complete();
+                }
                 CollectionAssert.AreEqual(
                     Enumerable.Repeat(module, 1),
                     registry.InstalledModules.Select(im => im.Module));
-
-                tScope.Complete();
             }
-            CollectionAssert.AreEqual(
-                Enumerable.Repeat(module, 1),
-                registry.InstalledModules.Select(im => im.Module));
         }
 
         [Test]
         public void TxRollback()
         {
-            // Our registry should roll-back any changes it made during a transaction.
-            // This one makes sure that aborting the transaction rolls back the change
-            var registry = CKAN.Registry.Empty();
-
-            using (var gameInstWrapper = new DisposableKSP())
-            using (var tScope = new TransactionScope())
+            var user = new NullUser();
+            using (var repoData = new TemporaryRepositoryData(user))
             {
-                var module = CkanModule.FromJson(@"{
-                                 ""spec_version"": ""v1.4"",
-                                 ""identifier"":   ""InstalledMod"",
-                                 ""author"":       ""InstalledModder"",
-                                 ""version"":      ""1.0"",
-                                 ""download"":     ""https://github.com/""
-                             }");
-                registry.RegisterModule(module, new List<string>(),
-                                        gameInstWrapper.KSP, false);
+                // Our registry should roll-back any changes it made during a transaction.
+                // This one makes sure that aborting the transaction rolls back the change
+                var registry = CKAN.Registry.Empty(repoData.Manager);
+
+                using (var gameInstWrapper = new DisposableKSP())
+                using (var tScope = new TransactionScope())
+                {
+                    var module = CkanModule.FromJson(@"{
+                                     ""spec_version"": ""v1.4"",
+                                     ""identifier"":   ""InstalledMod"",
+                                     ""author"":       ""InstalledModder"",
+                                     ""version"":      ""1.0"",
+                                     ""download"":     ""https://github.com/""
+                                 }");
+                    registry.RegisterModule(module, new List<string>(),
+                                            gameInstWrapper.KSP, false);
+
+                    CollectionAssert.AreEqual(
+                        Enumerable.Repeat(module, 1),
+                        registry.InstalledModules.Select(im => im.Module));
+
+                    // Rollback, our module should no longer be registered
+                    tScope.Dispose();
+                }
 
                 CollectionAssert.AreEqual(
-                    Enumerable.Repeat(module, 1),
+                    Enumerable.Empty<CkanModule>(),
                     registry.InstalledModules.Select(im => im.Module));
-
-                // Rollback, our module should no longer be registered
-                tScope.Dispose();
             }
-
-            CollectionAssert.AreEqual(
-                Enumerable.Empty<CkanModule>(),
-                registry.InstalledModules.Select(im => im.Module));
         }
 
         [Test]
         public void TxNested()
         {
-            // Our registry doesn't understand how to do nested transactions,
-            // make sure it throws on these.
-            // This one makes sure that one transaction inside another both work
-            // (except it doesn't check it? just that nothing throws?)
-            var registry = CKAN.Registry.Empty();
-
-            using (var gameInstWrapper = new DisposableKSP())
-            using (var tScope = new TransactionScope())
+            var user = new NullUser();
+            using (var repoData = new TemporaryRepositoryData(user))
             {
-                var module = CkanModule.FromJson(@"{
-                                 ""spec_version"": ""v1.4"",
-                                 ""identifier"":   ""InstalledMod"",
-                                 ""author"":       ""InstalledModder"",
-                                 ""version"":      ""1.0"",
-                                 ""download"":     ""https://github.com/""
-                             }");
-                registry.RegisterModule(module, new List<string>(),
-                                        gameInstWrapper.KSP, false);
+                // Our registry doesn't understand how to do nested transactions,
+                // make sure it throws on these.
+                // This one makes sure that one transaction inside another both work
+                // (except it doesn't check it? just that nothing throws?)
+                var registry = CKAN.Registry.Empty(repoData.Manager);
 
-                using (var tScope2 = new TransactionScope(TransactionScopeOption.RequiresNew))
+                using (var gameInstWrapper = new DisposableKSP())
+                using (var tScope = new TransactionScope())
                 {
-                    Assert.Throws<TransactionalKraken>(delegate
+                    var module = CkanModule.FromJson(@"{
+                                     ""spec_version"": ""v1.4"",
+                                     ""identifier"":   ""InstalledMod"",
+                                     ""author"":       ""InstalledModder"",
+                                     ""version"":      ""1.0"",
+                                     ""download"":     ""https://github.com/""
+                                 }");
+                    registry.RegisterModule(module, new List<string>(),
+                                            gameInstWrapper.KSP, false);
+
+                    using (var tScope2 = new TransactionScope(TransactionScopeOption.RequiresNew))
                     {
-                        var module2 = CkanModule.FromJson(@"{
-                                          ""spec_version"": ""v1.4"",
-                                          ""identifier"":   ""InstalledMod2"",
-                                          ""author"":       ""InstalledModder"",
-                                          ""version"":      ""1.0"",
-                                          ""download"":     ""https://github.com/""
-                                      }");
-                        registry.RegisterModule(module2, new List<string>(),
-                                                gameInstWrapper.KSP, false);
-                    });
-                    tScope2.Complete();
+                        Assert.Throws<TransactionalKraken>(delegate
+                        {
+                            var module2 = CkanModule.FromJson(@"{
+                                              ""spec_version"": ""v1.4"",
+                                              ""identifier"":   ""InstalledMod2"",
+                                              ""author"":       ""InstalledModder"",
+                                              ""version"":      ""1.0"",
+                                              ""download"":     ""https://github.com/""
+                                          }");
+                            registry.RegisterModule(module2, new List<string>(),
+                                                    gameInstWrapper.KSP, false);
+                        });
+                        tScope2.Complete();
+                    }
+                    tScope.Complete();
                 }
-                tScope.Complete();
             }
         }
 
         [Test]
         public void TxAmbient()
         {
-            // Our registry should be fine with ambient transactions, which join together.
-            // Note the absence of TransactionScopeOption.RequiresNew
-            var registry = CKAN.Registry.Empty();
-
-            using (var gameInstWrapper = new DisposableKSP())
-            using (var tScope = new TransactionScope())
+            var user = new NullUser();
+            using (var repoData = new TemporaryRepositoryData(user))
             {
-                var module = CkanModule.FromJson(@"{
-                                 ""spec_version"": ""v1.4"",
-                                 ""identifier"":   ""InstalledMod"",
-                                 ""author"":       ""InstalledModder"",
-                                 ""version"":      ""1.0"",
-                                 ""download"":     ""https://github.com/""
-                             }");
-                registry.RegisterModule(module, new List<string>(),
-                                        gameInstWrapper.KSP, false);
+                // Our registry should be fine with ambient transactions, which join together.
+                // Note the absence of TransactionScopeOption.RequiresNew
+                var registry = CKAN.Registry.Empty(repoData.Manager);
 
-                using (var tScope2 = new TransactionScope())
+                using (var gameInstWrapper = new DisposableKSP())
+                using (var tScope = new TransactionScope())
                 {
-                    Assert.DoesNotThrow(delegate
+                    var module = CkanModule.FromJson(@"{
+                                     ""spec_version"": ""v1.4"",
+                                     ""identifier"":   ""InstalledMod"",
+                                     ""author"":       ""InstalledModder"",
+                                     ""version"":      ""1.0"",
+                                     ""download"":     ""https://github.com/""
+                                 }");
+                    registry.RegisterModule(module, new List<string>(),
+                                            gameInstWrapper.KSP, false);
+
+                    using (var tScope2 = new TransactionScope())
                     {
-                        var module2 = CkanModule.FromJson(@"{
-                                          ""spec_version"": ""v1.4"",
-                                          ""identifier"":   ""InstalledMod2"",
-                                          ""author"":       ""InstalledModder"",
-                                          ""version"":      ""1.0"",
-                                          ""download"":     ""https://github.com/""
-                                      }");
-                        registry.RegisterModule(module2, new List<string>(),
-                                                gameInstWrapper.KSP, false);
-                    });
-                    tScope2.Complete();
+                        Assert.DoesNotThrow(delegate
+                        {
+                            var module2 = CkanModule.FromJson(@"{
+                                              ""spec_version"": ""v1.4"",
+                                              ""identifier"":   ""InstalledMod2"",
+                                              ""author"":       ""InstalledModder"",
+                                              ""version"":      ""1.0"",
+                                              ""download"":     ""https://github.com/""
+                                          }");
+                            registry.RegisterModule(module2, new List<string>(),
+                                                    gameInstWrapper.KSP, false);
+                        });
+                        tScope2.Complete();
+                    }
+                    tScope.Complete();
                 }
-                tScope.Complete();
             }
         }
     }

--- a/Tests/Core/Registry/RegistryManager.cs
+++ b/Tests/Core/Registry/RegistryManager.cs
@@ -79,7 +79,9 @@ namespace Tests.Core.Registry
                 var lockPath = Path.Combine(gameInst.KSP.CkanDir(), "registry.locked");
 
                 // Act / Assert
-                RegistryManager.Instance(gameInst.KSP, repoData.Manager);
+                var mgr = RegistryManager.Instance(gameInst.KSP, repoData.Manager);
+                Assert.IsFalse(File.Exists(jsonPath));
+                mgr.Save();
                 Assert.IsTrue(File.Exists(jsonPath));
                 Assert.IsTrue(File.Exists(lockPath));
                 RegistryManager.DisposeInstance(gameInst.KSP);

--- a/Tests/Core/Relationships/RelationshipResolverTests.cs
+++ b/Tests/Core/Relationships/RelationshipResolverTests.cs
@@ -39,10 +39,14 @@ namespace Tests.Core.Relationships
         [Test]
         public void Constructor_WithoutModules_AlwaysReturns()
         {
-            var registry = CKAN.Registry.Empty();
-            options = RelationshipResolverOptions.DefaultOpts(stabilityTolerance);
-            Assert.DoesNotThrow(() => new RelationshipResolver(new List<CkanModule>(),
-                null, options, registry, game, crit));
+            var user = new NullUser();
+            using (var repoData = new TemporaryRepositoryData(user))
+            {
+                var registry = CKAN.Registry.Empty(repoData.Manager);
+                options = RelationshipResolverOptions.DefaultOpts(stabilityTolerance);
+                Assert.DoesNotThrow(() => new RelationshipResolver(new List<CkanModule>(),
+                    null, options, registry, game, crit));
+            }
         }
 
         [Test]
@@ -1043,9 +1047,11 @@ namespace Tests.Core.Relationships
         [Test]
         public void AutodetectedCanSatisfyRelationships()
         {
+            var user = new NullUser();
+            using (var repoData = new TemporaryRepositoryData(user))
             using (var ksp = new DisposableKSP())
             {
-                var registry = CKAN.Registry.Empty();
+                var registry = CKAN.Registry.Empty(repoData.Manager);
                 registry.SetDlls(new Dictionary<string, string>()
                 {
                     {

--- a/Tests/GUI/Model/ModList.cs
+++ b/Tests/GUI/Model/ModList.cs
@@ -30,9 +30,11 @@ namespace Tests.GUI
         public void ComputeFullChangeSetFromUserChangeSet_WithEmptyList_HasEmptyChangeSet()
         {
             var item = new ModList();
+            var user = new NullUser();
+            using (var repoData = new TemporaryRepositoryData(user))
             using (var tidy = new DisposableKSP())
             {
-                Assert.That(item.ComputeUserChangeSet(Registry.Empty(), crit, tidy.KSP, null, null), Is.Empty);
+                Assert.That(item.ComputeUserChangeSet(Registry.Empty(repoData.Manager), crit, tidy.KSP, null, null), Is.Empty);
             }
         }
 
@@ -225,7 +227,7 @@ namespace Tests.GUI
                     {
                         // Install the "other" module
                         installer.InstallList(
-                            modList.ComputeUserChangeSet(Registry.Empty(), crit, inst2.KSP, null, null).Select(change => change.Mod).ToList(),
+                            modList.ComputeUserChangeSet(Registry.Empty(repoData.Manager), crit, inst2.KSP, null, null).Select(change => change.Mod).ToList(),
                             new RelationshipResolverOptions(inst2.KSP.StabilityToleranceConfig),
                             registryManager,
                             ref possibleConfigOnlyDirs,

